### ISSUE TITLE
Every job recompiles from scratch, and the same corpus was built eight times in two weeks

### ISSUE THREAD

**Issue by thodel** [thodel/serving-atr-inference#109]
#39 built chunked materialize → compile → discard and was closed on 2026-08-10. It
solved peak disk. It did not touch reuse, and reuse is where the time goes.

**A job id is the unit of everything.** `prepare` and `compile` write into
`jobs/<job-id>/`, so a new submission always starts from the hub, even when the
selection is byte-for-byte the one compiled ten minutes earlier. There is no way
to point a job at an existing `.arrow`.

## What that cost, measured

The same four-dataset German corpus — 12,300 pages, 325,700 lines, ~41 GB of
arrow — was compiled **eight times** between 24 August and 5 September:

| job | why it was rebuilt |
|---|---|
| `20260824T093013Z-kraken-medieval-german-v2` | OOM in train at `batch_size: 64` |
| `20260824T142549Z-…` | OOM at 32 |
| `20260825T043258Z-…` | OOM from one 135:1 line (#90) |
| `20260825T092433Z-…` | OOM at 16 |
| `20260825T145122Z-…` | ran; cancelled at epoch 44 |
| `20260905T185730Z-kraken-german-eval-pool-v1` | train failed at `batch_size: 256` |
| `20260905T190759Z-…` | **recompiled 10 minutes later, same spec, same failure** |
| `20260908T101611Z-qwen3vl-german-pages-v1` | page granularity, running |

Five of those differ from their predecessor **only in a hyperparameter** —
`batch_size`, mostly — and every one paid ~2.5 hours of prepare and compile to
change a number the train stage reads. Roughly **20 hours of wall clock and 330 GB
of writes** to produce artefacts that were already on disk.

The 5 September pair makes it starkest: two jobs, twelve minutes apart, identical
selection, identical 12,301 pages and 325,768 lines, and the second rebuilt all
41 GB before failing in exactly the same place as the first.

## Why it bites harder than it looks

- **A batch-size sweep is unaffordable.** Finding the value that fits took four
  runs at 2.5 h of setup each; the actual answer came from measuring the line-width
  distribution instead, because iterating was too expensive to iterate.
- **It fills the share.** 221 GB of dead arrows were deleted on 2026-09-08 and
  another 81 GB on 09-09, on a volume that was at 91 %.
- **It punishes exactly the runs that most need retrying** — the large ones.

## Scope

- [ ] A content key for a compiled artefact: `(hf_repo, revision, sorted projects,
      split, granularity, partition, seed)`. Two specs with the same key produce
      the same arrow, and that is checkable without reading the data.
- [ ] A cache keyed by it, outside `jobs/<job-id>/` — the job directory is the
      wrong home for something meant to outlive the job.
- [ ] `compile` consults it before building, and a job records which artefact it
      used, so a result stays traceable to its inputs.
- [ ] Eviction, since these are 40 GB each. LRU with a size budget, or explicit —
      but decided rather than left to fill the share again.

## Deliberately not in scope

Resuming a *partially* compiled artefact. That is the harder half (#39's chunk
boundaries were never made resumable either), and full-artefact reuse already
recovers most of the loss above: seven of the eight rebuilds were complete
compiles of a corpus that had finished compiling before.

## Note

The cache must key on the *spec*, not the job id, and must include `revision`.
Two jobs naming the same projects on different days can legitimately differ if the
dataset moved, and a cache that ignores that would serve stale pages while
reporting a fresh compile — worse than the waste it replaces.

**Comment by thodel** [comment]
Implemented in `afcf650` + `02fb9c9`. Pushed to `main`; **the box is not yet running it** — see the bottom of this comment.

## What landed

`src/atr_serving/training/artefact_cache.py` — the content key and the store. A job whose selection has been compiled before skips **prepare and compile together**: reusing the arrow while still downloading the pages that produced it would have saved the ~20 minutes of `ketos compile` and none of the two hours of prepare.

| in the key | not in the key |
|---|---|
| `hf_repo`, `revision`, `split` | `epochs`, `batch_size`, `lrate`, `spec` |
| `train_projects`, `eval_projects` (as **sets**) | `base_model`, `resize` |
| `partition`, `seed`, `max_pages`, `granularity` | `model_id`, `device`, `force` |
| the engine, and `chunk_pages` | everything else train reads |

Project order inside one dataset does not matter — a selection is a set. Dataset order does: a multi-dataset run pools pages with a per-dataset index offset, so reordering changes which page gets which index and therefore how the seeded split falls.

**kraken only.** A ketos binary dataset embeds the line images it was compiled from, so an `.arrow` reads from anywhere. The VLM backend's JSONL samples name image paths inside the job directory and do not survive being moved; it stays out until that is addressed.

## Three things worth stating

**An unpinned revision expires after 7 days.** `revision: null` means "whatever the dataset is today", and a cache that ignored that would serve last month's pages while reporting a fresh compile — worse than the waste it replaces. A spec that names a revision is reusable indefinitely.

**Eviction never touches an entry used in the last 72 hours.** Nothing tracks which job holds which artefact and a kraken run reads its arrow for the whole of training, so going over budget beats deleting the corpus out from under a run that is three days into it. When that happens the log says so.

**Skipping prepare deletes what the two guards measure.** The line and page counts and the geometry measurement travel with the artefact. That matters for the VGSL spec in particular: it is a *train* parameter, so it is not in the key, so a reused corpus can arrive under a spec the geometry guard has something to say about.

Every failure path — unreadable manifest, missing arrow, a filesystem that refuses the write — logs and compiles instead. A run that fails because of the cache would be strictly worse than one that was slow.

## One thing the box's own layout caught

`jobs_root` is on the CIFS share and the cache is on the system disk. The first cut staged the arrows into a directory beside the job and moved that to the cache — 41 GB over SMB **twice**, and the staging step could not use hard links because CIFS refuses them. `02fb9c9` hands the files to the cache directly: one copy, to the destination filesystem. Budget lowered 150 → 100 GB, because 457 GB free on the system disk is a real constraint rather than a formality.

## Operating it

```bash
python scripts/artefact_cache.py                 # what is in there
python scripts/artefact_cache.py --evict         # apply the budget now
python scripts/artefact_cache.py --drop 95b7c717 # remove one, by key prefix
```

Off with `ATR_TRAIN_ARTEFACT_CACHE=false`. Documented in `docs/TRAINING.md` §8b-bis.

Suite: **961 passed, 1 skipped**, verified on the staged tree in a clean worktree. `tests/conftest.py` redirects the cache root for the whole suite — the first run of these tests left five directories in `~/atr-cache/artefacts`, which is a bug in the suite rather than a quirk of it.

## Not deployed

The box's checkout is at `afcf650` and `atr-train` has **not** been restarted, so none of this is live. Remaining:

```bash
ssh tobias@130.92.59.240 'cd ~/Repo/serving-atr-inference && git pull --ff-only && systemctl --user restart atr-train'
```

Nothing is training in the service right now (`20260908T101611Z-qwen3vl-german-pages-v1` OOMed at 33%), and the `sweep_seed43.sh` run on the GPU is not a child of `atr-train`, so the restart will not disturb it.

**Comment by thodel** [comment]
Die Store-Seite dieses Caches kostet auf UBELIX 14 Stunden (966.748 Dateien, 87,5 GB, einzeln kopiert auf GPFS) — verfolgt in #148.

**Comment by thodel** [comment]
Closing: done, checked against the code on 17.09.2026. All four scope items are done. Kraken reuse landed in afcf650 and 02fb9c9. b06f8f5 (15.09) added the VLM backend. 24ec2c5 and 61ff968 raised the key to KEY_VERSION 3 and folded in the held-out registry digest. training-atr-models has the same file, differing only in the package name (src/atr_training/artefact_cache.py, KEY_VERSION=3 at :88). The hook is at runner_base.py:913 (_reuse_artefact), with _cache_key in the kraken and vlm runners. It is live on asteraix: training-atr-models docs/INFRASTRUCTURE.md:416 records 38 GiB on 16.09.2026 against a 100 GB budget. #148 was filed so that #109 could be closed ('umgesetzt, noch offen').
